src/operator/tensor/diag_op-inl.h

```diff
@@ -69,7 +82,28 @@ inline TShape DiagShapeImpl(const TShape& ishape, const nnvm::dim_t k) {
     s = 0;
   }
 
-  return TShape({s});
+  auto x1 = CheckAxis(axis1, ishape.ndim());
+  auto x2 = CheckAxis(axis2, ishape.ndim());
+
+  CHECK_NE(x1, x2) << "axis1 and axis2 cannot refer the the same axis " << x1;
+
+  if (x1 > x2) {
+    std::swap(x1, x2);
+  }
+
+  int n_dim = static_cast<int>(ishape.ndim()) - 1;
+  TShape oshape(n_dim);
+
+  // remove axis1 and axis2 and append the new axis to the end
+  for (int i = 0; i < x1; i ++)
+    oshape[i] = ishape[i];
+  for (int i = x1 + 1; i < x2; i ++)
+    oshape[i - 1] = ishape[i];
+  for (int i = x2 + 1; i <= n_dim; i ++)
+    oshape[i - 2] = ishape[i];
+  oshape[n_dim - 1] = s;
```

Comment:
Could this be simplified to something like:

int idx = 0;
for(int i=0; i<ndim; i++) {
   if(i != axis1 && i != axis2)
      oshape[idx++] = ishape[i]
oshape[ndim-2] = s;

Also, the final dimension assignment on line 104 looks wrong. Shouldnt it be oshape[n_dim - 2] = s;? Since the output tensor will be 1 dimension smaller than the input tensor (and off-by-1 for 0-based versus ndim being the count).

Reply:
Yes, the code could be simplified that way, but that does not necessarily make improvements in my opinion. The original implementation is equally easy to read and fairly efficient. As for the final assignment, note that `n_dim` is already the dimension of the input tensor minus 1, so I think it is fine.